### ISSUE TITLE
Plugin resolves dependencies at configuration time causing significant time loss for large multi module projects

### ISSUE THREAD

**Issue by itsyaboiyeah** [augi/gradle-wartremover#30]
https://github.com/augi/gradle-wartremover/blob/master/src/main/groovy/cz/augi/gradle/wartremover/WartremoverPlugin.groovy#L14

Resolves the dependencies at configuration time, as a consequence in our large multi module build wartremover was adding 20 secs to every configuration.

Can we work around this? I think it ought to be fine to defer the scalaTask.scalaCompileOptions.additionalParameters.add to a doFirst ?

**Comment by augi** [comment]
Hello, and thanks for your report! 👍 

Calling the resolving in `doFirst` could work but maybe it would be slightly better (~Gradlish) to call the resolving in `configure` of each affected task.

It would be great if you could test how both solutions works in your case.

**Comment by itsyaboiyeah** [comment]
Hello, thanks for the prompt response. 

I have tested doFirst, so that the above file changes to something like scalaTask.doFirst { ... 

And was able to update the tests to really run a gradle build using GradleRunner to see it still worked, it also works appropraitely in our codebase, still providing the warnings but now with no configuration overhead.

I'm not sure what you mean by resolving in configure of each task, I am happy to test it out, but if you could point me in the right direction on what this means that would be super.

**Comment by augi** [comment]
Which warnings did you see?

By the `configure`, I meant [configure](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskProvider.html#configure-org.gradle.api.Action-) method of `TaskProvider` - it is part of a bigger Gradle feature - [Task Configuration Avoidance API](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html). We are already using the right overload of `withType` method so maybe it could be sufficient to wrap the action in `configure` block.

**Comment by itsyaboiyeah** [comment]
Sorry, I was being confusing. The warnings I was referring to were the warnings from wartremover, my meaning was that after my proposed change, the functionality of wartremover still worked.

I was unable to make it work using configure, but being as I was unfamiliar with it, perhaps I am using it wrong. I have made a very small (the only content in a single build.gradle) repo, if you would like to clone that and try your proposed change.

https://github.com/itsyaboiyeah/resolution-sample

With the present wart remover, it correctly identifies that dependencies were resolved during configuration. If I then replace the WartremoverPlugin.groovy file with my proposedChange.groovy and do a --include-build, the dependencies are not resolved during configuration.
